modules/fledge.js

```diff
@@ -0,0 +1,192 @@
+  return componentAuctionsForAdUnitCode[randomIdx].fledgeAuctionConfig;
+}
+
+export function renderAdHook(next, doc, id, options) {
+  logInfo(MODULE, 'renderAd hook', arguments);
+  if (doc && id) {
+    try {
+      const bid = auctionManager.findBidByAdId(id);
+
+      if (bid) {
+        logInfo(MODULE, 'renderAd hook found bid', bid);
+        if (runAdAuction(bid, next, doc, id, options)) {
+          return;
+        }
+      }
+    } catch (e) {
+      logError(MODULE, 'fail running fledge', id, e);
+    }
+  }
+  next(doc, id, options);
+}
+
+export function runAdAuction(bid, next, doc, id, options) {
+  if (bid.mediaType == 'video') {
+    logWarn(MODULE, 'Video fledge ads not yet supported.');
+    // need to fallback to default renderAd
+    return false;
+  }
+
+  const auctionConfig = getAuctionConfig(bid);
+  if (auctionConfig) {
+    navigator.runAdAuction(auctionConfig).then((fledgeAdUri) => {
+      if (fledgeAdUri) {
+        renderFledgeAd(bid, doc, fledgeAdUri);
+      } else {
+        // need to fallback to default renderAd when there was no fledge ad
+        next(doc, id, options);
+      }
+    });
+    return true;
+  } else {
+    // need to fallback to default renderAd when there was no fledge auction config
+    return false;
+  }
+};
+
+export function renderFledgeAd(bid, doc, fledgeAdUri) {
+  logInfo(MODULE, `renderAd rendering fledge ad=${fledgeAdUri}`);
+
+  // TODO emit fledge win event
+
+  const {height, width} = bid;
+
+  // TODO switch to fencedframe when available
+  const frameType = 'iframe';
+  logInfo(MODULE, `renderAd in ${frameType}`, fledgeAdUri, doc);
+  const iframe = createInvisibleIframe();
+  iframe.style.display = 'inline';
+  iframe.style.overflow = 'hidden';
+  iframe.height = height;
+  iframe.width = width;
+  iframe.src = fledgeAdUri;
+  insertElement(iframe, doc, 'body');
+
+  if (doc.defaultView && doc.defaultView.frameElement) {
+    doc.defaultView.frameElement.width = width;
+    doc.defaultView.frameElement.height = height;
+  }
+
+  // Do billing/ reporting stuff here?
+  logInfo(MODULE, 'renderAd rendering fledge ad');
+}
+
+function secureRenderRequestHook(next, reply, data, adObject, options) {
+  logWarn('Fledge rendering using secureCreative');
+  const MODULE = 'fledge secure';
+  const fledgeAuctionConfig = getAuctionConfig(adObject);
+  if (fledgeAuctionConfig) {
+    logWarn(MODULE, `Ad id ${adObject.adId} is being fledged`, fledgeAuctionConfig);
+    navigator.runAdAuction(fledgeAuctionConfig).then((fledgeAdUri) => {
+      if (fledgeAdUri) {
+        logWarn(MODULE, `Ad id ${adObject.adId} has been fledged by ${fledgeAdUri}`);
+        adObject.ad = '';
+        adObject.adUrl = fledgeAdUri;
+      } else {
+        logWarn(MODULE, `Ad id ${adObject.adId} was NOT fledged, rendering normal ad`);
+      }
+      // we use the default secure rendering
+      next(reply, data, adObject, {isFledge: true});
```

Comment:
I think `isFledge` here should be true only if you got a `fledgeAdUri` back from `navigator.runAdAuction`.